### PR TITLE
Add deferrable mode to Databricks SQL warehouse operators

### PR DESCRIPTION
**Pull request by Vamsi-klu**
This adds an additive deferrable path for the Databricks SQL warehouse start and stop operators, so lifecycle waits can run on the triggerer instead of holding a worker.

related: #70088
related: #21377

### Problem

Airflow's Databricks provider can start and stop an existing SQL warehouse, but Phase 1 waits hold a worker for the entire poll. Dag authors who want triggerer-based waiting currently have no `deferrable` path on these operators.

### What changed

- Add `DatabricksHook.a_get_warehouse_state` as the async GET mirror of `get_warehouse_state`.
- Restore `WarehouseState.to_json` / `from_json` for trigger event round-trip.
- Add `DatabricksWarehouseStateTrigger` next to the provider's existing triggers, with serialize, local monotonic timeout, retry_args validation, and a documented no-op `on_kill`.
- Add `deferrable` on both `DatabricksStartWarehouseOperator` and `DatabricksStopWarehouseOperator` using the `default_deferrable` pattern.
- Document the deferrable path in the warehouse how-to and add a provider changelog Features note.

No request-body workaround or new dependency is introduced. The implementation follows the [Databricks SQL Warehouses API](https://docs.databricks.com/api/workspace/warehouses) and keeps the Phase 1 wait contract.

### Scope

This is the Phase 2 deferrable follow-up agreed on #70088: async hook, serialized trigger, operator completion path, and compatibility tests. Create/delete, edit, warehouse-by-name resolution, and a live Databricks system-test Dag remain outside this PR so the follow-up stays reviewable and independently useful.

### Behavior and compatibility

- `deferrable` defaults off and honors `[operators] default_deferrable`.
- `wait_for_termination`, `polling_period_seconds`, and `timeout` are unchanged.
- Starting an already `RUNNING` warehouse and stopping an already `STOPPED` warehouse remain no-ops and do not defer.
- `wait_for_termination=False` still returns after requesting the transition and does not defer.
- Existing `STARTING`/`STOPPING` transitions are reused instead of issuing duplicate requests, then deferred if waiting.
- A start accepted while the warehouse still reports `STOPPING` continues on the triggerer; Databricks API transition rejections propagate unchanged.
- Stale `STOPPED` while waiting for `RUNNING` is not terminal; the trigger keeps polling until `RUNNING`, deletion, or timeout.
- Waiting uses `time.monotonic()` locally in the trigger, starts no new poll after the configured deadline, and still honors a target or deletion state returned by a poll that began before the deadline.
- Clearing a deferred wait does not start or stop the warehouse: Databricks has no cancel API for these transitions.
- `execute_complete` maps `success` / `deleted` / `timeout` to the same messages as Phase 1.

### Validation

- `uv run --project providers/databricks pytest providers/databricks/tests/unit/databricks/operators/test_warehouse.py providers/databricks/tests/unit/databricks/hooks/test_databricks.py::TestWarehouseLifecycle providers/databricks/tests/unit/databricks/triggers/test_databricks.py::TestDatabricksWarehouseStateTrigger providers/databricks/tests/unit/databricks/triggers/test_databricks.py::test_trigger_init_rejects_non_serializable_retry_args` — 70 passed.
- `prek run --from-ref upstream/main --stage pre-commit` — passed.

### Reviewer evidence

This PR has no UI surface, so before/after screenshots and browser validation are not applicable. The REST boundary is covered with an autospecced async GET assertion, operator deferral is covered with a specced hook, and trigger serialize/run/timeout/on_kill are covered with mocked warehouse state. No Databricks workspace credentials were used or required for these deterministic lifecycle tests.

The Phase 2 scope was posted on #70088 before this follow-up: https://github.com/apache/airflow/pull/70088#issuecomment-5235779087

---

##### Was generative AI tooling used to co-author this PR?

- [X] Yes — Cursor Grok 4.6

Generated-by: Cursor Grok 4.6 following [the guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#gen-ai-assisted-contributions)

### CODE REVIEW

**Comment by Vamsi-klu** [comment]
@eladkal @moomindani ni this is the #70088 deferrable follow-up, could you review when you have a chance?

File: providers/databricks/docs/changelog.rst [hunk 1]
```diff
@@ -29,6 +29,11 @@ Changelog
 7.18.1
 ......
 
+Features
+~~~~~~~~
+
+* ``Add deferrable mode to Databricks SQL warehouse start and stop operators``
```
Comment (moomindani) [inline providers/databricks/docs/changelog.rst]: This block sits under the `7.18.1` header, which the release manager created for the 2026-08-06 wave — a released version, so its changelog would advertise a feature that release does not contain, and it will collide with the entry regenerated from `git log` for the next version.

`providers/AGENTS.md` and the `NOTE TO CONTRIBUTORS` block at the top of this file both say routine feature entries are collected automatically and direct edits are reserved for breaking/important behaviour notes. Simplest fix is to drop all five lines; the commit message already carries the note.

File: providers/databricks/src/airflow/providers/databricks/triggers/databricks.py [hunk 2]
```diff
@@ -336,3 +336,128 @@ async def run(self):
...
+            "target_state": self.target_state,
+            "last_state": last_state,
+        }
+        if state is not None:
+            payload["state"] = state.to_json()
+        return TriggerEvent(payload)
+
+    async def run(self):
+        async with self.hook:
+            deadline = time.monotonic() + self.timeout
```
Comment (moomindani) [inline providers/databricks/src/airflow/providers/databricks/triggers/databricks.py]: The deadline is computed here, inside `run()`, and `serialize()` carries `timeout` as a duration — so after a triggerer restart, deploy, or HA rebalance the wait starts a fresh full `timeout` window, and a warehouse that never reaches its target can stay deferred well past the configured value. The synchronous path fails deterministically instead.

`DatabricksSQLStatementExecutionTrigger` in this same file takes an absolute `end_time` (serialized at line 281, consumed at line 299) that the operator computes, precisely so the deadline survives serialization. Following that pattern here would make the two consistent; if the per-run semantics are deliberate instead, they are worth stating in the docstring and the how-to.

Related: the docstring above says the monotonic deadline "survives trigger serialization", and the PR description repeats it. It is the other way round — a monotonic value cannot be serialized, which is why this line recomputes it.

File: providers/databricks/src/airflow/providers/databricks/triggers/databricks.py [hunk 3]
```diff
@@ -336,3 +336,128 @@ async def run(self):
...
+                "timeout": self.timeout,
+                "polling_period_seconds": self.polling_period_seconds,
+                "retry_limit": self.retry_limit,
+                "retry_delay": self.retry_delay,
+                "retry_args": self.retry_args,
+                "caller": self.caller,
+            },
+        )
+
+    async def on_kill(self) -> None:
```
Comment (moomindani) [inline providers/databricks/src/airflow/providers/databricks/triggers/databricks.py]: `BaseTrigger.on_kill` is already a documented no-op, so this override only adds a log line. I deleted it entirely and `test_on_kill_leaves_warehouse_unchanged` still passed, which means the test does not pin the behaviour the description credits it with.

Either drop both the override and the test and keep the reasoning as a comment, or keep the override and stop describing it as tested. Worth noting the base-class method only exists on newer cores while the provider supports `apache-airflow>=2.11.0`.

File: providers/databricks/src/airflow/providers/databricks/hooks/databricks.py [hunk 4]
```diff
@@ -799,6 +806,15 @@ def get_warehouse_state(self, warehouse_id: str) -> WarehouseState:
         """
         return WarehouseState(self.get_warehouse(warehouse_id)["state"])
 
+    async def a_get_warehouse_state(self, warehouse_id: str) -> WarehouseState:
```
Comment (moomindani) [inline providers/databricks/src/airflow/providers/databricks/hooks/databricks.py:817]: This inlines the endpoint tuple, while the synchronous `get_warehouse_state` goes through `get_warehouse()`. They are identical today — same `2.0` constant, same `WarehouseState(response["state"])` parsing — but that is now two copies, so a later change to `get_warehouse` (query params, error mapping, an API version bump) would silently miss the async path. An `a_get_warehouse` mirror, or a small shared endpoint builder, keeps them in step.

File: providers/databricks/src/airflow/providers/databricks/triggers/databricks.py [hunk 5]
```diff
@@ -336,3 +336,128 @@ async def run(self):
...
+        self, *, status: str, last_state: str, state: WarehouseState | None = None
+    ) -> TriggerEvent:
+        payload: dict[str, Any] = {
+            "status": status,
+            "warehouse_id": self.warehouse_id,
+            "target_state": self.target_state,
+            "last_state": last_state,
+        }
+        if state is not None:
+            payload["state"] = state.to_json()
```
Comment (moomindani) [inline providers/databricks/src/airflow/providers/databricks/triggers/databricks.py:424]: `payload["state"]` is always `json.dumps({"state": payload["last_state"]})`, and `execute_complete` then does `WarehouseState.from_json(event["state"]).state` where `event["last_state"]` would do — while hard-indexing `event["state"]` and `.get()`-ing everything else. Worth either dropping the redundant key or keeping it deliberately for symmetry with the sibling triggers and saying so.

On the `to_json`/`from_json` restore: this closes the loop from #70088, where their removal was requested because nothing used them and you noted they would return with this trigger. That is the strongest justification for the change and the description does not mention it.

**Review by moomindani — CHANGES_REQUESTED** [review]
Thanks — the Phase 1 contract is preserved carefully here, and the state-machine edge cases we went through on #70088 (stale `STOPPED` not terminal, deletion mid-wait, the deadline boundary) are all covered by deterministic tests rather than sleeps.

I validated the deferrable paths against a real workspace: Airflow 3.2.2, this branch's provider installed from source, a dedicated throwaway serverless warehouse (2X-Small), triggerer doing the waiting.

| scenario | task state | evidence |
| --- | --- | --- |
| deferrable start from `STOPPED` | deferred -> success | trigger logged `is STARTING; waiting for RUNNING`, warehouse reached `RUNNING` |
| deferrable stop | deferred -> success | warehouse reached `STOPPED` |
| start against an already-`RUNNING` warehouse | success, never deferred | no `DeferTask` for the task |
| `wait_for_termination=False` stop | success, never deferred | transition still requested, warehouse stopped afterwards |
| `timeout=1` start | deferred -> failed | `DatabricksWarehouseError: ... did not reach RUNNING within 1s; last state: STARTING.` |

So the behavioural claims in the description hold up where I could exercise them. Unit tests reproduce your numbers as well (43 + 14 + 13, and the full provider suite at 922 passed / 12 skipped).

Two things I would like settled before merge, both left inline:

- The changelog entry is inserted under the `7.18.1` header, which is an already-released version, so a shipped release would advertise a feature it does not contain. Per `providers/AGENTS.md` and the `NOTE TO CONTRIBUTORS` block at the top of that file, routine feature entries are collected by the release manager from commit messages anyway — the five lines can just go.
- The deferrable `timeout` is a duration recomputed inside every `run()`, so it restarts from zero on each triggerer restart or HA rebalance, while the synchronous path fails deterministically. `DatabricksSQLStatementExecutionTrigger` next door serializes an absolute `end_time` for exactly this reason. Either follow that pattern or state the per-run semantics explicitly — right now the docstring and the PR description claim the opposite of what the code does.

The rest is nits, also inline. One out-of-scope note worth being aware of rather than fixing here: these operators' first deferrable surface inherits #71525, so on core 3.0.x the trigger cannot fetch the connection at all (3.1.0+ is fine). Whether that caveat belongs in this page or centrally is a call for the docs discussion happening on #71667.

File: providers/databricks/docs/changelog.rst (hunk 1 above)
Comment (Vamsi-klu) [reply providers/databricks/docs/changelog.rst]: Dropped the Features block under 7.18.1. The commit message already carries the note for the release manager.

File: providers/databricks/src/airflow/providers/databricks/triggers/databricks.py (hunk 2 above)
Comment (Vamsi-klu) [reply providers/databricks/src/airflow/providers/databricks/triggers/databricks.py]: Followed DatabricksSQLStatementExecutionTrigger. The operator now records an absolute end_time from timeout when it defers, serialize carries that timestamp, and run() compares it with time.time(). A triggerer restart no longer resets the wait. The how-to and the trigger docstring now describe that.

File: providers/databricks/src/airflow/providers/databricks/triggers/databricks.py (hunk 3 above)
Comment (Vamsi-klu) [reply providers/databricks/src/airflow/providers/databricks/triggers/databricks.py]: Removed the on_kill override and the test that only asserted the base no-op. The class docstring now says we leave warehouse state unchanged because Databricks has no cancel API for start or stop.

File: providers/databricks/src/airflow/providers/databricks/hooks/databricks.py (hunk 4 above)
Comment (Vamsi-klu) [reply providers/databricks/src/airflow/providers/databricks/hooks/databricks.py:817]: Added a_get_warehouse as the async mirror of get_warehouse. a_get_warehouse_state now goes through that method so the GET path stays in one place.

File: providers/databricks/src/airflow/providers/databricks/triggers/databricks.py (hunk 5 above)
Comment (Vamsi-klu) [reply providers/databricks/src/airflow/providers/databricks/triggers/databricks.py:424]: Kept the serialized state key so the payload matches the RunState and SQLStatementState triggers, which is also why WarehouseState.to_json and from_json came back after the Phase 1 removal. last_state stays as the plain string used in the timeout message. The PR description now calls out that restore from 70088.

**Comment by Vamsi-klu** [comment]
> Thanks — the Phase 1 contract is preserved carefully here, and the state-machine edge cases we went through on #70088 (stale `STOPPED` not terminal, deletion mid-wait, the deadline boundary) are all covered by deterministic tests rather than sleeps.
> 
> I validated the deferrable paths against a real workspace: Airflow 3.2.2, this branch's provider installed from source, a dedicated throwaway serverless warehouse (2X-Small), triggerer doing the waiting.
> 
> | scenario | task state | evidence |
> | --- | --- | --- |
> | deferrable start from `STOPPED` | deferred -> success | trigger logged `is STARTING; waiting for RUNNING`, warehouse reached `RUNNING` |
> | deferrable stop | deferred -> success | warehouse reached `STOPPED` |
> | start against an already-`RUNNING` warehouse | success, never deferred | no `DeferTask` for the task |
> | `wait_for_termination=False` stop | success, never deferred | transition still requested, warehouse stopped afterwards |
> | `timeout=1` start | deferred -> failed | `DatabricksWarehouseError: ... did not reach RUNNING within 1s; last state: STARTING.` |
> 
> So the behavioural claims in the description hold up where I could exercise them. Unit tests reproduce your numbers as well (43 + 14 + 13, and the full provider suite at 922 passed / 12 skipped).
> 
> Two things I would like settled before merge, both left inline:
> 
> - The changelog entry is inserted under the `7.18.1` header, which is an already-released version, so a shipped release would advertise a feature it does not contain. Per `providers/AGENTS.md` and the `NOTE TO CONTRIBUTORS` block at the top of that file, routine feature entries are collected by the release manager from commit messages anyway — the five lines can just go.
> - The deferrable `timeout` is a duration recomputed inside every `run()`, so it restarts from zero on each triggerer restart or HA rebalance, while the synchronous path fails deterministically. `DatabricksSQLStatementExecutionTrigger` next door serializes an absolute `end_time` for exactly this reason. Either follow that pattern or state the per-run semantics explicitly — right now the docstring and the PR description claim the opposite of what the code does.
> 
> The rest is nits, also inline. One out-of-scope note worth being aware of rather than fixing here: these operators' first deferrable surface inherits #71525, so on core 3.0.x the trigger cannot fetch the connection at all (3.1.0+ is fine). Whether that caveat belongs in this page or centrally is a call for the docs discussion happening on #71667.
> 

Thanks for the live workspace check @moomindani i responded to comments and commited the latest change

The follow-up is on the branch. Changelog Features under 7.18.1 is gone. The trigger now serializes an absolute end_time the same way DatabricksSQLStatementExecutionTrigger does. a_get_warehouse shares the GET path with get_warehouse. The on_kill override and the unpinned test are gone. The typed state JSON stays for sibling trigger symmetry, and the description now mentions the Phase 1 to_json restore.

The 3.0.x connection fetch caveat from 71525 is still out of scope here.

**Review by moomindani — APPROVED** [review]
Thanks — all five points are addressed, and I re-verified on `38bc535d4a` rather than taking it on trust.

The changelog is back to the released `7.18.1` section untouched. The deadline is now an absolute `end_time` computed by the operator, serialized, and compared with `time.time()` in `run()`, matching `DatabricksSQLStatementExecutionTrigger` — and the two new tests pin it where it matters: the round trip preserves `end_time`, and a trigger resumed with an already-passed `end_time` times out immediately instead of opening a fresh window. That was the part I cared about most. `on_kill` and its vacuous test are gone with the rationale kept in the class docstring, and `a_get_warehouse_state` now goes through `a_get_warehouse`; the remaining duplicated path suffix matches how `start_warehouse` / `stop_warehouse` declare theirs in that file, so no objection from me. Keeping the typed state payload for sibling symmetry with the reasoning recorded is a fine call.

Real workspace re-run on this head (Airflow 3.2.2, provider installed from the branch, dedicated throwaway serverless warehouse): deferrable start `deferred -> success`, deferrable stop `deferred -> success`, and `timeout=1` `deferred -> failed` with `did not reach RUNNING within 1s`. 203 unit tests pass across the three files.

One observation from that last run, not a request: with a `timeout` shorter than the defer round-trip, the absolute deadline has already passed by the time the triggerer picks the trigger up, so it never polls and the message reports `last state: unknown` where the synchronous path would have polled once and named a state. That is the correct consequence of honouring an absolute deadline and is invisible at realistic timeouts — just noting that the boundary now differs slightly between the two paths.

LGTM.

---
Drafted-by: Claude Code (Opus 5); reviewed by @moomindani before posting

**Comment by Vamsi-klu** [comment]
@eladkal as @moomindani approved the PR, Can i get your thoughts on this please?

**Comment by eladkal** [comment]
can you please update the comment in https://github.com/apache/airflow/issues/21377#issuecomment-5014011201 with checklist to the actual steps in the plan and which PR solved which step? That way we can easily track what is left open

**Comment by Vamsi-klu** [comment]
Hi @eladkal thanks a ton for the comment

Updated the tracking comment: https://github.com/apache/airflow/issues/21377#issuecomment-5014011201

#70088 is Phase 1. This PR is the deferrable follow-up. Create/delete, name lookup, edit, and a live system-test Dag are still open.

**Review by eladkal — APPROVED** [review]
